plugins/chan_track.py

```diff
@@ -266,87 +458,112 @@ def perm_check(chan, conn, nick):
 def cleanusers(bot):
+    """
+    :type bot: cloudbot.bot.CloudBot
+    """
     clean_data(bot)
     gc.collect()
     return "Data cleaned."
 
 
 @hook.command(permissions=["botcontrol"], autohelp=False)
 def clearusers(bot):
-    init_chan_data(bot, True)
+    """
+    :type bot: cloudbot.bot.CloudBot
+    """
+    for conn in bot.connections.values():
+        init_chan_data(conn, True)
+
     gc.collect()
     return "Data cleared."
 
 
+@hook.command("getdata", permissions=["botcontrol"], autohelp=False)
+def getdata_cmd(conn, chan, text, nick):
+    chan_data = get_chans(conn).getchan(chan)
+    user_data = get_users(conn).getuser(nick)
+    memb = chan_data.get_member(user_data)
+    return web.paste(MappingSerializer().serialize(memb, indent=2))
+
+
 @hook.irc_raw('JOIN')
 def on_join(nick, user, host, conn, irc_paramlist):
+    """
+    :type nick: str
+    :type user: str
+    :type host: str
+    :type conn: cloudbot.client.Client
+    :type irc_paramlist: cloudbot.util.parsers.irc.ParamList
+    """
     chan, *other_data = irc_paramlist
 
     if chan.startswith(':'):
         chan = chan[1:]
 
-    data = {'ident': user, 'host': host}
+    users = get_users(conn)
 
-    if is_cap_available(conn, "extended-join") and other_data:
-        acct, realname = other_data
-        if acct == "*":
-            acct = None
-
-        data.update(account=acct, realname=realname)
+    user_data = users.getuser(nick)
 
-    users = conn.memory['users']
+    user_data.ident = user
+    user_data.host = host
 
-    user_data = users.getuser(nick)
-    user_data.update(data)
+    if is_cap_available(conn, "extended-join") and other_data:
+        acct, realname = other_data
+        user_data.account = acct
+        user_data.realname = realname
 
-    chan_data = conn.memory["chan_data"].getchan(chan)
-    memb_data = WeakDict(chan=weakref.proxy(chan_data), user=user_data, status=[])
-    chan_data["users"][nick] = memb_data
-    add_user_membership(user_data, chan, memb_data)
+    chan_data = get_chans(conn).getchan(chan)
+    user_data.join_channel(chan_data)
 
 
 @hook.irc_raw('MODE')
 def on_mode(chan, irc_paramlist, conn):
+    """
+    :type chan: str
+    :type conn: cloudbot.client.Client
+    :type irc_paramlist: cloudbot.util.parsers.irc.ParamList
+    """
     if chan.startswith(':'):
         chan = chan[1:]
 
+    if chan.casefold() == conn.nick.casefold():
+        # this is a user mode line
+        return
+
     serv_info = conn.memory["server_info"]
     statuses = serv_info["statuses"]
     status_modes = {status.mode for status in statuses.values()}
     mode_types = serv_info["channel_modes"]
-    chans = conn.memory["chan_data"]
 
-    try:
-        chan_data = chans[chan]
-    except KeyError:
-        return
+    chan_data = get_chans(conn).getchan(chan)
 
-    chan_users = chan_data["users"]
     modes = irc_paramlist[1]
     mode_params = irc_paramlist[2:]
     new_modes = {}
```

Comment:
This is far too much branching for a single function. I'd suggest either breaking the logic out into separate functions or refactoring it altogether.